File: ext/internal/stats/metrics.go
package stats

import (
	"os"
	"runtime/metrics"
	"strconv"
	"strings"
	"time"
)

// clockTicksPerSecond is USER_HZ — the unit of the utime/stime fields in
// /proc/self/stat. It is fixed at 100 on virtually every Linux build (the kernel
// default), so we use it directly rather than pulling in cgo for sysconf.
const clockTicksPerSecond = 100.0

// goRuntimeMetric is the runtime/metrics name for all memory the Go runtime has
// mapped read-write — the equivalent of runtime.MemStats.Sys, but read without a
// stop-the-world (ReadMemStats stops the world; metrics.Read does not).
const goRuntimeMetric = "/memory/classes/total:bytes"

// readMemory builds the process memory split: the whole resident set from the OS
// (/proc/self/status VmRSS), the Go runtime's own footprint (runtime/metrics), and
// the remainder attributed to the non-extension side (PHP interpreter).
func readMemory() Memory {
	rssBytes := readRssBytes()

	goRuntimeBytes := readGoRuntimeBytes()

	nonExtensionBytes := rssBytes - goRuntimeBytes

	if nonExtensionBytes < 0 {
		nonExtensionBytes = 0
	}

	return Memory{
		RssBytes:          rssBytes,
		GoRuntimeBytes:    goRuntimeBytes,
		NonExtensionBytes: nonExtensionBytes,
	}
}

// readGoRuntimeBytes reads the Go runtime's total mapped memory via runtime/metrics,
// which (unlike runtime.ReadMemStats) does not stop the world — so sampling it on the
// hot push cadence does not pause the worker's request handlers.
func readGoRuntimeBytes() int64 {
	samples := []metrics.Sample{{Name: goRuntimeMetric}}

	metrics.Read(samples)

	if samples[0].Value.Kind() != metrics.KindUint64 {
		return 0
	}

	return int64(samples[0].Value.Uint64())
}

// readRssBytes reads the resident set size from /proc/self/status (the VmRSS
// line, reported in kB). Returns 0 if the value cannot be read.
func readRssBytes() int64 {
	contents, err := os.ReadFile("/proc/self/status")

	if err != nil {
		return 0
	}

	for _, line := range strings.Split(string(contents), "\n") {
		if !strings.HasPrefix(line, "VmRSS:") {
			continue
		}

		fields := strings.Fields(line)

		if len(fields) < 2 {
			return 0
		}

		kiloBytes, err := strconv.ParseInt(fields[1], 10, 64)

		if err != nil {
			return 0
		}

		return kiloBytes * 1024
	}

	return 0
}

// cpuSampler turns the monotonically growing CPU time of the process into a
// rolling percentage: each call diffs the consumed CPU ticks against wall time
// since the previous call. The first call only seeds the baseline and returns 0.
type cpuSampler struct {
	previousTicks float64
	previousWall  time.Time
	seeded        bool
}

// sample returns the CPU usage percent over the interval since the previous call.
// now is passed in so the writer loop uses one consistent clock for CPU and the
// snapshot timestamp.
func (sampler *cpuSampler) sample(now time.Time) float64 {
	ticks := readProcessCpuTicks()

	if !sampler.seeded {
		sampler.seeded = true
		sampler.previousTicks = ticks
		sampler.previousWall = now

		return 0
	}

	deltaTicks := ticks - sampler.previousTicks
	deltaWallSeconds := now.Sub(sampler.previousWall).Seconds()

	sampler.previousTicks = ticks
	sampler.previousWall = now

	if deltaWallSeconds <= 0 {
		return 0
	}

	cpuSeconds := deltaTicks / clockTicksPerSecond

	return cpuSeconds / deltaWallSeconds * 100.0
}

// readProcessCpuTicks reads utime+stime (in clock ticks) from /proc/self/stat.
// The comm field (2nd) may contain spaces and parentheses, so we parse the fields
// after the last ')': there state is field 3, making utime field 14 (index 11)
// and stime field 15 (index 12). Returns 0 on any parse failure.
func readProcessCpuTicks() float64 {
	contents, err := os.ReadFile("/proc/self/stat")

	if err != nil {
		return 0
	}

	text := string(contents)

	lastParen := strings.LastIndex(text, ")")

	if lastParen < 0 || lastParen+2 > len(text) {
		return 0
	}

	fields := strings.Fields(text[lastParen+2:])

	// index 11 = utime (field 14), index 12 = stime (field 15).
	if len(fields) < 13 {
		return 0
	}

	utime, errUtime := strconv.ParseFloat(fields[11], 64)
	stime, errStime := strconv.ParseFloat(fields[12], 64)

	if errUtime != nil || errStime != nil {
		return 0
	}

	return utime + stime
}

File: ext/src/stats/metrics.rs
//! The process metrics every snapshot carries, read from /proc rather than from
//! any runtime API — so they measure the whole worker, PHP included, and not
//! just this half of it.

use std::time::Instant;

use super::Memory;

/// USER_HZ — the unit of the utime/stime fields in /proc/self/stat. Fixed at 100
/// on virtually every Linux build (the kernel default), so it is used directly
/// rather than reaching for `sysconf`.
const CLOCK_TICKS_PER_SECOND: f64 = 100.0;

/// The worker's resident set size, PHP and extension together.
///
/// There is deliberately no split between the two. Attributing a resident page
/// to one side would need a tracking global allocator, which is an atomic on
/// every allocation this core makes; the fields that used to carry the split
/// reported a zero and a copy of the total, which is worse than not reporting
/// it.
pub fn read_memory() -> Memory {
    Memory {
        rss_bytes: read_rss_bytes(),
    }
}

/// Reads the resident set size from /proc/self/status (the VmRSS line, in kB).
/// Answers 0 when it cannot be read.
fn read_rss_bytes() -> i64 {
    let Ok(contents) = std::fs::read_to_string("/proc/self/status") else {
        return 0;
    };

    for line in contents.lines() {
        let Some(rest) = line.strip_prefix("VmRSS:") else {
            continue;
        };

        return rest
            .split_whitespace()
            .next()
            .and_then(|value| value.parse::<i64>().ok())
            .map(|kilobytes| kilobytes * 1024)
            .unwrap_or(0);
    }

    0
}

/// Turns the monotonically growing CPU time of the process into a rolling
/// percentage: each sample diffs the consumed CPU ticks against wall time since
/// the previous one. The first sample only seeds the baseline and answers 0.
pub struct CpuSampler {
    previous: Option<(f64, Instant)>,
}

impl CpuSampler {
    pub fn new() -> Self {
        CpuSampler { previous: None }
    }

    /// `now` is passed in so one clock reading covers both the CPU percentage and
    /// the snapshot's own timestamp.
    pub fn sample(&mut self, now: Instant) -> f64 {
        let ticks = read_process_cpu_ticks();

        let Some((previous_ticks, previous_wall)) = self.previous.replace((ticks, now)) else {
            return 0.0;
        };

        let wall_seconds = now.duration_since(previous_wall).as_secs_f64();

        if wall_seconds <= 0.0 {
            return 0.0;
        }

        (ticks - previous_ticks) / CLOCK_TICKS_PER_SECOND / wall_seconds * 100.0
    }
}

/// Reads utime+stime (in clock ticks) from /proc/self/stat.
///
/// The comm field (the 2nd) may hold spaces and parentheses, so the fields are
/// counted from after the last ')': state is then field 3, which puts utime at
/// field 14 (index 11) and stime at field 15 (index 12). Answers 0 on any parse
/// failure.
fn read_process_cpu_ticks() -> f64 {
    let Ok(contents) = std::fs::read_to_string("/proc/self/stat") else {
        return 0.0;
    };

    let Some(last_paren) = contents.rfind(')') else {
        return 0.0;
    };

    let Some(tail) = contents.get(last_paren + 1..) else {
        return 0.0;
    };

    let fields: Vec<&str> = tail.split_whitespace().collect();

    if fields.len() < 13 {
        return 0.0;
    }

    // The state field (the 3rd of the line) is index 0 of this tail, so utime and
    // stime — fields 14 and 15 — land at indexes 11 and 12.
    let utime = fields[11].parse::<f64>();
    let stime = fields[12].parse::<f64>();

    match (utime, stime) {
        (Ok(utime), Ok(stime)) => utime + stime,
        _ => 0.0,
    }
}
